state/executor.go

```diff
@@ -123,6 +129,82 @@ type BlockResult struct {
 	TotalGas uint64
 }
 
+// getStakingEventType is a helper method for getting the type of system event
+func getStakingEventType(toAddress types.Address) staking.StakingEventType {
+	if toAddress == types.StringToAddress(system.StakingAddress) {
+		return staking.StakingEvent
+	}
+
+	return staking.UnstakingEvent
+}
+
+// constructPendingEvent is a helper method for constructing a pending event
+func constructPendingEvent(
+	toAddress types.Address,
+	fromAddress types.Address,
+	value *big.Int,
+	blockNumber uint64,
+) staking.PendingEvent {
+	eventType := getStakingEventType(toAddress)
+
+	return staking.PendingEvent{
+		Number:    int64(blockNumber),
+		Address:   fromAddress,
+		Value:     value,
+		EventType: eventType,
+	}
+}
+
+// cleanDiscardedEvents is a helper method for clearing discarded pending events
+func cleanDiscardedEvents(transactions []*types.Transaction, blockNumber uint64) {
+	for _, discardedTxn := range transactions {
+		// Remove a pending event for the discarded transaction
+		if system.IsSystemEvent(discardedTxn.To) {
+			pendingEvent := constructPendingEvent(
+				*discardedTxn.To,
+				discardedTxn.From,
+				discardedTxn.Value,
+				blockNumber,
+			)
+
+			hub := staking.GetStakingHub()
+			if hub.ContainsPendingEvent(pendingEvent) {
+				// Remove the event from the queue
+				hub.RemovePendingEvent(pendingEvent)
+			}
```

Comment:
We traverse EventQueue twice here. I think we can call only RemovePendingEvent here, or RemovePendingEvent can return `deleted` flag for other code (it's useful the code in commitApprovedEvents)

Reply:
You're right, there is no need to traverse it twice. Removing the double traversal 👍